File: cracking_the_coding_interview/3-6.cpp
#include <iostream>
#include <queue>
#include <chrono>
#include <cassert>
using namespace std;

enum AnimalKind{ DOG, CAT };

struct AnimalData{
    string name;
    AnimalKind kind;
};

class AnimalShelter{
public:
    void enqueue(AnimalData animal);
    AnimalData dequeueAny();
    AnimalData dequeueCat();
    AnimalData dequeueDog();

private:
    queue<pair<string, chrono::system_clock::time_point>> catQueue;
    queue<pair<string, chrono::system_clock::time_point>> dogQueue;
};

void AnimalShelter::enqueue(AnimalData animal){
    auto now = std::chrono::system_clock::now();
    if(animal.kind == AnimalKind::CAT){
        catQueue.push(make_pair(animal.name, now));
    }else{
        dogQueue.push(make_pair(animal.name, now));
    }
}

AnimalData AnimalShelter::dequeueAny(){
    if(catQueue.empty() && dogQueue.empty()){
        AnimalData animal;
        return animal;
    }

    if(dogQueue.empty()){
        return dequeueCat();
    }
    if(catQueue.empty()){
        return dequeueDog();
    }
    if(catQueue.front().second < dogQueue.front().second){
        return dequeueCat();
    }
    return dequeueDog();
}

AnimalData AnimalShelter::dequeueCat(){
    AnimalData animal;
    if(!catQueue.empty()){
        animal.kind = AnimalKind::CAT;
        animal.name = catQueue.front().first;
        catQueue.pop();
    }
    return animal;
}

AnimalData AnimalShelter::dequeueDog(){
    AnimalData animal;
    if(!dogQueue.empty()){
        animal.kind = AnimalKind::DOG;
        animal.name = dogQueue.front().first;
        dogQueue.pop();
    }
    return animal;
}

int main(void){
    AnimalData cat1 = {"tama", AnimalKind::CAT};
    AnimalData cat2 = {"mike", AnimalKind::CAT};
    AnimalData dog1 = {"pochi", AnimalKind::DOG};
    AnimalData dog2 = {"shiro", AnimalKind::DOG};
    AnimalShelter animalShelter;
    animalShelter.enqueue(cat1);
    animalShelter.enqueue(cat2);
    animalShelter.enqueue(dog1);
    animalShelter.enqueue(dog2);
    AnimalData output1 = animalShelter.dequeueAny(); // cat1
    AnimalData output2 = animalShelter.dequeueDog(); // dog1
    AnimalData output3 = animalShelter.dequeueAny(); // cat2
    AnimalData output4 = animalShelter.dequeueCat(); // none
    AnimalData output5 = animalShelter.dequeueAny(); // dog2
    cout << output1.name << endl;
    assert(output1.name == cat1.name);
    cout << output2.name << endl;
    assert(output2.name == dog1.name);
    cout << output3.name << endl;
    assert(output3.name == cat2.name);
    // AnimalData animal;
    // assert(output4 == animal);
    cout << output5.name << endl;
    assert(output5.name == dog2.name);
    return 0;
}

File: cracking_the_coding_interview/3-6.py
from enum import Enum
import time

class Animal(Enum):
    DOG = 0
    CAT = 1

class ListNode(object):
    def __init__(self, name, breed, time, next=None):
        self.name = name
        self.breed = breed
        self.time = time
        self.next = next

class AnimalShelter(object):
    def __init__(self):
        self.dog_head = ListNode("", Animal.DOG, 0)
        self.cat_head = ListNode("", Animal.CAT, 0)
        self.dog_last = self.dog_head
        self.cat_last = self.cat_head
    
    def enqueue(self, name, breed):
        t = time.time()
        node = ListNode(name, breed, t)
        if breed == Animal.DOG:
            self.dog_last.next = node
            self.dog_last = node
        else:
            self.cat_last.next = node
            self.cat_last = node
    
    def dequeue_any(self):
        if not self.dog_head.next:
            return self.dequeue_cat()
        if not self.cat_head.next:
            return self.dequeue_dog()
        
        dog_t = self.dog_head.next.time
        cat_t = self.cat_head.next.time
        return self.dequeue_dog() if dog_t < cat_t else self.dequeue_cat()
    
    def dequeue_dog(self):
        if not self.dog_head.next:
            return None
        node = self.dog_head.next
        self.dog_head.next = self.dog_head.next.next
        return node
    
    def dequeue_cat(self):
        if not self.cat_head.next:
            return None
        node = self.cat_head.next
        self.cat_head.next = self.cat_head.next.next
        return node

if __name__ == '__main__':
    al = AnimalShelter()
    al.enqueue("Pochi", Animal.DOG)
    al.enqueue("Mike", Animal.CAT)
    al.enqueue("Shiro", Animal.DOG)
    al.enqueue("Kotetsu", Animal.DOG)
    al.enqueue("Tama", Animal.CAT)
    print(al.dequeue_dog().name)
    print(al.dequeue_any().name)
    print(al.dequeue_cat().name)
    print(al.dequeue_cat().name)
    print(al.dequeue_any().name)

